### PR TITLE
feat(server)!: pgvecto.rs 0.2 and pgvector compatibility

### LINKED ISSUE #5813: [BUG] Don't work immich (1.91.0(include this version) next version)
### The bug

i work immich on docker with synology DS1815+
i did worked until 1.90.2.
but, update 1.91.0 over(1.91.0 ~ 1.91.3), i didnt work my immich.
(and, i did tried all remove and new install by 1.91.3)

logs of immich_postgres... repeat this error code..

QueryFailedError: called `Result::unwrap()` on an `Err` value: Os { code: 38, kind: Unsupported, message: "Function not implemented" }
    at async DataSource.initialize (/usr/src/app/node_modules/typeorm/data-source/DataSource.js:148:17)
    at PostgresQueryRunner.query (/usr/src/app/node_modules/typeorm/driver/postgres/PostgresQueryRunner.js:211:19)
    at async AddCLIPEmbeddingIndex1700713994428.up (/usr/src/app/dist/infra/migrations/1700713994428-AddCLIPEmbeddingIndex.js:9:9)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async MigrationExecutor.executePendingMigrations (/usr/src/app/node_modules/typeorm/migration/MigrationExecutor.js:225:17)
    at async DataSource.runMigrations (/usr/src/app/node_modules/typeorm/data-source/DataSource.js:260:35)

and, logs of immich_serverm, repeat this error
[Nest] 9  - 12/18/2023, 4:32:00 AM   ERROR [TypeOrmModule] Unable to connect to the database. Retrying (6)...
QueryFailedError: called `Result::unwrap()` on an `Err` value: Os { code: 38, kind: Unsupported, message: "Function not implemented" }
    at PostgresQueryRunner.query (/usr/src/app/node_modules/typeorm/driver/postgres/PostgresQueryRunner.js:211:19)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async AddCLIPEmbeddingIndex1700713994428.up (/usr/src/app/dist/infra/migrations/1700713994428-AddCLIPEmbeddingIndex.js:9:9)
    at async MigrationExecutor.executePendingMigrations (/usr/src/app/node_modules/typeorm/migration/MigrationExecutor.js:225:17)
    at async DataSource.runMigrations (/usr/src/app/node_modules/typeorm/data-source/DataSource.js:260:35)
    at async DataSource.initialize (/usr/src/app/node_modules/typeorm/data-source/DataSource.js:148:17)
plz, help me...TT

### The OS that Immich Server is running on

Synology 7.1.1-42962 update 6.(uname -a result : Linux RANSER-NAS-BOX 3.10.108 #42962 SMP Mon May 29 14:36:28 CST 2023 x86_64 GNU/Linux synology_avoton_1815+)

### Version of Immich Server

v1.91.3

### Version of Immich Mobile App

v1.91.3

### Platform with the issue

- [X] Server
- [X] Web
- [ ] Mobile

### Your docker-compose.yml content

```YAML
version: "3.8"

#
# WARNING: Make sure to use the docker-compose.yml of the current release:
#
# https://github.com/immich-app/immich/releases/latest/download/docker-compose.yml
#
# The compose file on main may not be compatible with the latest release.
#

name: tmimmich

services:
  immich-server:
    container_name: immich_server
    image: ghcr.io/immich-app/immich-server:${IMMICH_VERSION:-release}
    command: [ "start.sh", "immich" ]
    volumes:
      - ${UPLOAD_LOCATION}:/usr/src/app/upload
      - ${RAWDATA_LOCATION}:/usr/src/app/upload/library
      - /etc/localtime:/etc/localtime:ro
      - ${EXTERNAL_LOCATION}:/mnt/media/old_photo
    env_file:
      - .env
    ports:
      - 2283:3001
    depends_on:
      - redis
      - database
    restart: always
    networks:
      TMBRIDGE:
        ipv4_address: 172.18.1.101

  immich-microservices:
    container_name: immich_microservices
    image: ghcr.io/immich-app/immich-server:${IMMICH_VERSION:-release}
    # extends:
    #   file: hwaccel.yml
    #   service: hwaccel
    command: [ "start.sh", "microservices" ]
    volumes:
      - ${UPLOAD_LOCATION}:/usr/src/app/upload
      - ${RAWDATA_LOCATION}:/usr/src/app/upload/library
      - /etc/localtime:/etc/localtime:ro
      - ${EXTERNAL_LOCATION}:/mnt/media/old_photo
    env_file:
      - .env
    depends_on:
      - redis
      - database
    restart: always
    networks:
      TMBRIDGE:
        ipv4_address: 172.18.1.102

  immich-machine-learning:
    container_name: immich_machine_learning
    image: ghcr.io/immich-app/immich-machine-learning:${IMMICH_VERSION:-release}
    volumes:
      - model-cache:/cache
      - /etc/localtime:/etc/localtime:ro
    env_file:
      - .env
    restart: always
    networks:
      TMBRIDGE:
        ipv4_address: 172.18.1.103

  redis:
    container_name: immich_redis
    image: redis:6.2-alpine@sha256:80cc8518800438c684a53ed829c621c94afd1087aaeb59b0d4343ed3e7bcf6c5
    volumes:
      - /etc/localtime:/etc/localtime:ro
    restart: always
    networks:
      TMBRIDGE:
        ipv4_address: 172.18.1.105

  database:
    container_name: immich_postgres
    image: tensorchord/pgvecto-rs:pg14-v0.1.11
    env_file:
      - .env
    environment:
      POSTGRES_PASSWORD: ${DB_PASSWORD}
      POSTGRES_USER: ${DB_USERNAME}
      POSTGRES_DB: ${DB_DATABASE_NAME}
    volumes:
      - pgdata:/var/lib/postgresql/data
      - /etc/localtime:/etc/localtime:ro
    restart: always
    networks:
      TMBRIDGE:
        ipv4_address: 172.18.1.106

volumes:
  pgdata:
  model-cache:

networks:
  TMBRIDGE:
    external: true
```

### Your .env content

```Shell
You can find documentation for all the supported env variables at https://immich.app/docs/install/environment-variables

# The location where your uploaded files are stored
UPLOAD_LOCATION=./data
RAWDATA_LOCATION=/volume2/TMPHOTO/immich-photofile
EXTERNAL_LOCATION=/volume2/homes/TM/newphotos

# The Immich version to use. You can pin this to a specific version like "v1.71.0"
IMMICH_VERSION=release

# Connection secrets for postgres and typesense. You should change these to random passwords
DB_PASSWORD=REMOVE

# The values below this line do not need to be changed
###################################################################################
DB_HOSTNAME=immich_postgres
DB_USERNAME=postgres
DB_DATABASE_NAME=immich

REDIS_HOSTNAME=immich_redis
```

### Reproduction steps

```bash
1. docker-compose down 
2. docker-compose up -d
...
```

### Additional information
on immich_postgres log file
PostgreSQL Database directory appears to contain a database; Skipping initialization
2023-12-18 13:58:08.898 KST [1] LOG:  starting PostgreSQL 14.10 (Debian 14.10-1.pgdg120+1) on x86_64-pc-linux-gnu, compiled by gcc (Debian 12.2.0-14) 12.2.0, 64-bit
2023-12-18 13:58:08.899 KST [1] LOG:  listening on IPv4 address "0.0.0.0", port 5432
2023-12-18 13:58:08.899 KST [1] LOG:  listening on IPv6 address "::", port 5432
2023-12-18 13:58:08.914 KST [1] LOG:  listening on Unix socket "/var/run/postgresql/.s.PGSQL.5432"
2023-12-18 13:58:08.932 KST [27] LOG:  database system was shut down at 2023-12-18 13:57:29 KST
2023-12-18 13:58:08.956 KST [1] LOG:  database system is ready to accept connections
2023-12-18 13:58:23.451 KST [37] ERROR:  called `Result::unwrap()` on an `Err` value: Os { code: 38, kind: Unsupported, message: "Function not implemented" }
2023-12-18 13:58:23.451 KST [37] STATEMENT:  
	      CREATE INDEX IF NOT EXISTS clip_index ON smart_search
	      USING vectors (embedding cosine_ops) WITH (options = $$
	      [indexing.hnsw]
	      m = 16
	      ef_construction = 300
	      $$);
2023-12-18 13:58:24.580 KST [38] ERROR:  called `Result::unwrap()` on an `Err` value: Os { code: 38, kind: Unsupported, message: "Function not implemented" }
2023-12-18 13:58:24.580 KST [38] STATEMENT:  
	      CREATE INDEX IF NOT EXISTS clip_index ON smart_search
	      USING vectors (embedding cosine_ops) WITH (options = $$
	      [indexing.hnsw]
	      m = 16
	      ef_construction = 300
	      $$);

_No response_

### LINKED ISSUE #5880: [BUG] pgvecto.rs: IPC connection is closed unexpected.
### The bug

```
[Nest] 8  - 12/20/2023, 1:24:29 AM   ERROR [JobService] QueryFailedError: pgvecto.rs: pgvecto.rs: IPC connection is closed unexpected.
ADVICE: The error is raisen by background worker errors. Please check the full Postgresql log to get more information.
    at PostgresQueryRunner.query (/usr/src/app/node_modules/typeorm/driver/postgres/PostgresQueryRunner.js:211:19)
    at process.processTicksAndRejections (node:internal/process/task_queues:95:5)
    at async InsertQueryBuilder.execute (/usr/src/app/node_modules/typeorm/query-builder/InsertQueryBuilder.js:106:33)
    at async SmartInfoRepository.upsertEmbedding (/usr/src/app/dist/infra/repositories/smart-info.repository.js:110:9)
    at async SmartInfoRepository.upsert (/usr/src/app/dist/infra/repositories/smart-info.repository.js:103:9)
    at async SmartInfoService.handleEncodeClip (/usr/src/app/dist/domain/smart-info/smart-info.service.js:103:9)
    at async /usr/src/app/dist/domain/job/job.service.js:113:37
    at async Worker.processJob (/usr/src/app/node_modules/bullmq/dist/cjs/classes/worker.js:387:28)
    at async Worker.retryIfFailed (/usr/src/app/node_modules/bullmq/dist/cjs/classes/worker.js:574:24)
 ```

### The OS that Immich Server is running on

Kubernetes, 1.27

### Version of Immich Server

v1.91.4

### Version of Immich Mobile App

v1.0.0

### Platform with the issue

- [X] Server
- [ ] Web
- [ ] Mobile

### Your docker-compose.yml content

```YAML
apiVersion: apps/v1
kind: Deployment
metadata:
  annotations:
    deployment.kubernetes.io/revision: "8"
    meta.helm.sh/release-name: immich-microservices
    meta.helm.sh/release-namespace: apps
    reloader.stakater.com/auto: "true"
  creationTimestamp: "2023-11-07T23:53:01Z"
  generation: 9
  labels:
    app.kubernetes.io/component: main
    app.kubernetes.io/instance: immich-microservices
    app.kubernetes.io/managed-by: Helm
    app.kubernetes.io/name: immich-microservices
    helm.sh/chart: app-template-2.0.3
    helm.toolkit.fluxcd.io/name: immich-microservices
    helm.toolkit.fluxcd.io/namespace: apps
  name: immich-microservices
  namespace: apps
  resourceVersion: "277312664"
  uid: b1aa4cb1-9e7e-4e70-ba5a-c2010d7c5661
spec:
  progressDeadlineSeconds: 600
  replicas: 1
  revisionHistoryLimit: 3
  selector:
    matchLabels:
      app.kubernetes.io/component: main
      app.kubernetes.io/instance: immich-microservices
      app.kubernetes.io/name: immich-microservices
  strategy:
    rollingUpdate:
      maxSurge: 25%
      maxUnavailable: 25%
    type: RollingUpdate
  template:
    metadata:
      creationTimestamp: null
      labels:
        app.kubernetes.io/component: main
        app.kubernetes.io/instance: immich-microservices
        app.kubernetes.io/name: immich-microservices
    spec:
      automountServiceAccountToken: true
      containers:
      - args:
        - start-microservices.sh
        env:
        - name: DB_DATABASE_NAME
          valueFrom:
            secretKeyRef:
              key: dbname
              name: immich-pgsql-pguser-immich
        - name: DB_HOSTNAME
          valueFrom:
            secretKeyRef:
              key: host
              name: immich-pgsql-pguser-immich
        - name: DB_PASSWORD
          valueFrom:
            secretKeyRef:
              key: password
              name: immich-pgsql-pguser-immich
        - name: DB_PORT
          value: "5432"
        - name: DB_USERNAME
          valueFrom:
            secretKeyRef:
              key: user
              name: immich-pgsql-pguser-immich
        - name: DISABLE_REVERSE_GEOCODING
          value: "true"
        - name: IMMICH_MEDIA_LOCATION
          value: /usr/src/app/upload
        - name: LOG_LEVEL
          value: log
        - name: MICROSERVICES_PORT
          value: "3002"
        - name: REDIS_HOSTNAME
          valueFrom:
            configMapKeyRef:
              key: REDIS_HOSTNAME
              name: immich-configmap
        - name: REDIS_PORT
          valueFrom:
            configMapKeyRef:
              key: REDIS_PORT
              name: immich-configmap
        - name: STAKATER_IMMICH_CONFIGMAP_CONFIGMAP
          value: 0cd40a1ba7a4c36cbfd8a510784480b5a1973ea3
        image: ghcr.io/immich-app/immich-server:v1.91.4
        imagePullPolicy: IfNotPresent
        name: main
        resources:
          limits:
            memory: 8000Mi
          requests:
            cpu: 100m
            memory: 512Mi
        terminationMessagePath: /dev/termination-log
        terminationMessagePolicy: File
        volumeMounts:
        - mountPath: /usr/src/app/.reverse-geocoding-dump
          name: geocoding-dump
        - mountPath: /usr/src/app/node_modules/local-reverse-geocoder/geonames_dump
          name: geoname-dump
        - mountPath: /usr/src/app/upload
          name: library
      dnsPolicy: ClusterFirst
      enableServiceLinks: false
      restartPolicy: Always
      schedulerName: default-scheduler
      securityContext:
        fsGroup: 568
        fsGroupChangePolicy: OnRootMismatch
        runAsGroup: 568
        runAsUser: 568
        supplementalGroups:
        - 44
        - 109
        - 65542
      serviceAccount: default
      serviceAccountName: default
      terminationGracePeriodSeconds: 30
      topologySpreadConstraints:
      - labelSelector:
          matchLabels:
            app.kubernetes.io/name: immich-microservices
        maxSkew: 1
        topologyKey: kubernetes.io/hostname
        whenUnsatisfiable: DoNotSchedule
      volumes:
      - emptyDir: {}
        name: geocoding-dump
      - emptyDir: {}
        name: geoname-dump
      - name: library
        persistentVolumeClaim:
          claimName: immich-nfs
status:
  availableReplicas: 1
  conditions:
  - lastTransitionTime: "2023-11-07T23:53:01Z"
    lastUpdateTime: "2023-12-20T01:15:12Z"
    message: ReplicaSet "immich-microservices-68ccc9ccf4" has successfully progressed.
    reason: NewReplicaSetAvailable
    status: "True"
    type: Progressing
  - lastTransitionTime: "2023-12-20T01:23:56Z"
    lastUpdateTime: "2023-12-20T01:23:56Z"
    message: Deployment has minimum availability.
    reason: MinimumReplicasAvailable
    status: "True"
    type: Available
  observedGeneration: 9
  readyReplicas: 1
  replicas: 1
  updatedReplicas: 1

```
```


### Your .env content

```Shell
- name: DB_DATABASE_NAME
          valueFrom:
            secretKeyRef:
              key: dbname
              name: immich-pgsql-pguser-immich
        - name: DB_HOSTNAME
          valueFrom:
            secretKeyRef:
              key: host
              name: immich-pgsql-pguser-immich
        - name: DB_PASSWORD
          valueFrom:
            secretKeyRef:
              key: password
              name: immich-pgsql-pguser-immich
        - name: DB_PORT
          value: "5432"
        - name: DB_USERNAME
          valueFrom:
            secretKeyRef:
              key: user
              name: immich-pgsql-pguser-immich
        - name: DISABLE_REVERSE_GEOCODING
          value: "true"
        - name: IMMICH_MEDIA_LOCATION
          value: /usr/src/app/upload
        - name: LOG_LEVEL
          value: log
        - name: MICROSERVICES_PORT
          value: "3002"
        - name: REDIS_HOSTNAME
          valueFrom:
            configMapKeyRef:
              key: REDIS_HOSTNAME
              name: immich-configmap
        - name: REDIS_PORT
          valueFrom:
            configMapKeyRef:
              key: REDIS_PORT
              name: immich-configmap
```

### Reproduction steps

```bash
1. Run microservices, any task running is errored with  pgvecto.rs:
```

### Additional information

pgvecto.rs = 0.1.11

### LINKED ISSUE #5893: [BUG] pgvector.rs Kernel 3.10
### The bug

New Postgres database in its 0.1.11 release is not compatible with 3.10.108 kernel (present in most Synology NAS).
Issue seems to be fixed in pgvector.rs: https://github.com/tensorchord/pgvecto.rs/pull/204

### The OS that Immich Server is running on

DSM 7.2

### Version of Immich Server

v1.9.4

### Version of Immich Mobile App

v1.91.4

### Platform with the issue

- [X] Server
- [ ] Web
- [ ] Mobile

### Your docker-compose.yml content

```YAML
version: "3.9"

name: immich

services:
  immich-server:
    container_name: immich_server
    image: ghcr.io/immich-app/immich-server:${IMMICH_VERSION:-release}
    command: [ "start.sh", "immich" ]
    volumes:
      - ${UPLOAD_LOCATION}:/usr/src/app/upload
      - /volume1/Photos/albums:/mnt/media/albums:ro
      - /etc/localtime:/etc/localtime:ro
    env_file:
      - .env
    ports:
      - 2283:3001
    depends_on:
      - redis
      - database
    restart: always

  immich-microservices:
    container_name: immich_microservices
    image: ghcr.io/immich-app/immich-server:${IMMICH_VERSION:-release}
    # extends:
    #   file: hwaccel.yml
    #   service: hwaccel
    command: [ "start.sh", "microservices" ]
    volumes:
      - ${UPLOAD_LOCATION}:/usr/src/app/upload
      - /volume1/Photos/albums:/mnt/media/albums:ro
      - /etc/localtime:/etc/localtime:ro
    env_file:
      - .env
    depends_on:
      - redis
      - database
    restart: always

  immich-machine-learning:
    container_name: immich_machine_learning
    image: ghcr.io/immich-app/immich-machine-learning:${IMMICH_VERSION:-release}
    volumes:
      - ./model-cache:/cache
    env_file:
      - .env
    restart: always

  redis:
    container_name: immich_redis
    image: redis:6.2-alpine@sha256:60e49e22fa5706cd8df7d5e0bc50ee9bab7c608039fa653c4d961014237cca46
    restart: always

  database:
    container_name: immich_postgres
    image: tensorchord/pgvecto-rs:pg14-v0.1.11
    env_file:
      - .env
    environment:
      POSTGRES_PASSWORD: ${DB_PASSWORD}
      POSTGRES_USER: ${DB_USERNAME}
      POSTGRES_DB: ${DB_DATABASE_NAME}
    volumes:
      - ./pgdata:/var/lib/postgresql/data
    restart: always
```

### Your .env content

```Shell
# You can find documentation for all the supported env variables at https://immich.app/docs/install/environment-variables

# The location where your uploaded files are stored
UPLOAD_LOCATION=/volume1/Photos

# The Immich version to use. You can pin this to a specific version like "v1.71.0"
#IMMICH_VERSION="v1.90.2"

# Connection secrets for postgres and typesense. You should change these to random passwords
DB_PASSWORD=my-secret-password

# The values below this line do not need to be changed
###################################################################################
DB_HOSTNAME=immich_postgres
DB_USERNAME=postgres
DB_DATABASE_NAME=immich

REDIS_HOSTNAME=immich_redis
```

### Reproduction steps

```bash
1. Pull the the last release of Immich
2. immich_postgres provides errors: 

immich_postgres | 2023-12-19 01:46:11.483 UTC [42] ERROR: called Result::unwrap() on an Err value: Os { code: 38, kind: Unsupported, message: "Function not implemented" }
immich_postgres | 2023-12-19 01:46:11.483 UTC [42] STATEMENT:
immich_postgres | CREATE INDEX IF NOT EXISTS clip_index ON smart_search
immich_postgres | USING vectors (embedding cosine_ops) WITH (options = $$
immich_postgres | [indexing.hnsw]
immich_postgres | m = 16
immich_postgres | ef_construction = 300
immich_postgres | $$);
```

### Additional information

_No response_

### PR DESCRIPTION
**Pull request by mertalev**
## Description

This PR upgrades the codebase to be compatible with pgvecto.rs 0.2 and future patch releases. This change is beneficial in several respects:
* It allows for searching with VBASE, a novel vector search algorithm that allows fetching an arbitrary number of rows instead of a fixed number upfront
* There have been many bug fixes as well as stability and performance improvements since the 0.1.11 release
* It makes adding and maintaining compatibility with pgvector easier
* Updating is simpler for admins as they can use any 0.2.* release, not a specific patch release
* Maintaining Immich is easier for third-party platforms as they have more freedom in the pgvecto.rs version they use
* It fixes a kernel issue with some Synology servers

Additionally, it now allows pgvector to be used instead. This is added for compatibility reasons - there are environments where pgvecto.rs simply cannot be used right now.

Summary of changes:
* Check for available extension upgrades at startup and update if possible
* Generalize vector extension code to work with both pgvecto.rs and pgvector
  * Use pgvecto.rs's pgvector compatibility mode for more code reusability between the two
* Add an env that determines which extension is used
   * Migrations are applied based on this env, so this choice cannot be changed after a successful startup
* Make parts of database startup run in a lock to prevent both server and microservices from making changes at the same time
* Add `vectors` to the database search path
  * pgvecto.rs now uses this schema to be able to coexist with pgvector
* Update version constraints to allow pinning to the patch/minor/major part of the minimum version

To do:
* [x] Add more unit tests

Follow-up work:
* [ ] Refactor the vector DB code - there's some duplication and the logic is spread apart across the database and smart info repos as well as migrations
  * It might be good to move the existing dim size update and migration code all to the database repo
* [ ] Update e2e to run with both extensions to make sure we don't break pgvector
* [x] Prepare documentation
  * The instructions in the 1.91.0 release notes will be outdated - should we consider making a page for this in the docs?
* [x] Do more manual testing to discover possible edge-cases

Fixes #5813
Fixes #5880
Fixes #5893
Addresses discussion #5954

### CODE REVIEW

**Comment by OverHash** [comment]
Should also help with #5813

**Comment by mertalev** [comment]
> Should also help with #5813

Right, added that to the PR description

File: server/src/infra/1707000751533-AddVectorsToSearchPath.ts [hunk 1]
```diff
@@ -0,0 +1,11 @@
+import { MigrationInterface, QueryRunner } from 'typeorm';
+
+export class AddVectorsToSearchPath1707000751533 implements MigrationInterface {
+  public async up(queryRunner: QueryRunner): Promise<void> {
+    await queryRunner.query(`ALTER DATABASE immich SET search_path TO "$user", public, vectors`);
```
Comment (danieldietzler) [inline server/src/infra/1707000751533-AddVectorsToSearchPath.ts:5]: We allow a custom database name (see https://immich.app/docs/install/environment-variables#database) so we should probably use that env variable here as well?

**Review by danieldietzler — APPROVED** [review]
Nice! I really like those new error messages for wrong installs. That's a great addition and probably helps a lot reducing support tickets related to this.

File: server/src/infra/1707000751533-AddVectorsToSearchPath.ts (hunk 1 above)
Comment (jrasm91) [reply server/src/infra/1707000751533-AddVectorsToSearchPath.ts]: I believe there are other migrations that have the database name hardcoded to `immich` by the way.

Comment (waclaw66) [reply server/src/infra/1707000751533-AddVectorsToSearchPath.ts]: > I believe there are other migrations that have the database name hardcoded to `immich` by the way.

I don't think so, using `immich_dev` and haven't any problem yet. Haven't found any occurrence of `immich` db within migrations.

**Comment by ScrumpyJack** [comment]
Thank you so much for adding pgvector compat. Fantastic!
